src/plugins/intel_gna/src/gna_plugin.cpp

```diff
@@ -210,69 +234,44 @@ void GNAPlugin::ExportScores(void *ptr_dst,
                   uint32_t num_active_elements,
                   uint32_t num_vector_stride,
                   Precision precision_in,
-                  Precision precision_out) {
+                  Precision precision_out,
+                  const float scale_factor) {
+    OV_ITT_SCOPED_TASK(itt::domains::GNAPlugin, "ExportScores");
+
     if (ptr_src == nullptr || ptr_dst == nullptr) {
         THROW_GNA_EXCEPTION << "Received null pointer arguments";
     }
+
     if (precision_out != Precision::I32 && precision_out != Precision::FP32) {
         THROW_GNA_EXCEPTION << "Unsupported target precision for infer : " << precision_out.name();
     }
+
+    const auto precision_in_size = precision_in.size();
     // source scores are possibly padded to multiple of 8 and possibly interleaved
     // rotate if necessary and only copy actual scores (not padding)
     if (orientation == kDnnInterleavedOrientation) {
-        int32_t *dst = reinterpret_cast<int32_t *>(ptr_dst);
-        const int8_t *src = reinterpret_cast<const int8_t*>(ptr_src);
+        const uint8_t *src = reinterpret_cast<const uint8_t*>(ptr_src);
         for (uint32_t i = 0; i < num_frames; i++) {
             for (uint32_t j = 0; j < num_active_elements; j++) {
-                auto input_ptr = src + (j * num_group + i) * precision_in.size();
-                auto dst_ptr = dst + (i * num_vector_elements + j);
-
-                switch (precision_in) {
-                    case Precision::I8 : {
-                        *dst_ptr = static_cast<int32_t>(*reinterpret_cast<const int8_t*>(input_ptr));
-                        break;
-                    }
-                    case Precision::I16 : {
-                        *dst_ptr  = static_cast<int32_t>(*reinterpret_cast<const int16_t*>(input_ptr));
-                        break;
-                    }
-                    case Precision::I32 : {
-                        *dst_ptr = *reinterpret_cast<const int32_t *>(input_ptr);
-                        break;
-                    }
-                    default:
-                        THROW_GNA_EXCEPTION << "Unsupported output layer precision: " << precision_in.name();
-                }
+                auto dst_ptr = ptr_dst + (i * num_vector_elements + j);
+                auto input_ptr = src + (j * num_group + i) * precision_in_size;
+                UpscaleAndCast(dst_ptr, input_ptr, precision_in, scale_factor);
```

Comment:
IMO calling `UpscaleAndCast` in a loop may still lead to poor accuracy ...
The compiler may still use branches for switch/case on `precision_in` for every converted number.